MSC2385: Disable URL Previews, alternative method (#2385)
matrix-org/matrix-doc: proposals/2385-no-url-previews.md

# MSC 2385: Disable URL Previews
This MSC is an alternative suggestion to [MSC2376](https://github.com/matrix-org/matrix-doc/pull/2376).

Currently URLs posted in chat give a URL preview. Sometimes, for example with bots, this can be
unwanted. For instance if a bot is posting a message with a lot of links, or the content of the link
is already obvious from the message itself.

## Proposal
This proposal adds a new, optional key to `m.room.message` messages, named `url_previews`. It is an

> array of strings containing the URLs a preview should be given to. Is this key absent it is expected

Reviewer: ```suggestion
array of strings containing the URLs a preview should be given to. If this key absent it is expected
```

> array of strings containing the URLs a preview should be given to. Is this key absent it is expected

Reviewer 2: ```suggestion
array of strings containing the URLs a preview should be given to. If this key is absent, it is expected
```

> that the client gives URL previews to all present URLs. Is a URL present in this key that is *not*

Reviewer: ```suggestion
that the client gives URL previews for all URLs present in the message body. If a URL is present in this key that is *not*
```

> that the client gives URL previews for all URLs present in the message body. If a URL is present in this key that is *not*

Reviewer 3: We can't expect that clients give URL previews as this could be a security problem. We should say that clients MAY provide previews for all URLs in the body.

Author: Good point. Also FYI url previews are proxied through your homeserver, so you don't actually have any security problem when generating those.

Reviewer 4: However, there is a problem: WWW URLs may return different things depending on other log-ins and the like, so the sender and the homeserver of the sender may not see the same content.

Author: https://matrix.org/docs/spec/client_server/r0.6.1#get-matrix-media-r0-preview-url discussion on that is beyond this MSC. The top comment suggests to use "may" instead of "it is expected", which soru will incorporate into this MSC

Reviewer 3: But it reveals the URL to the home server. You can imagine a case where a malicious user sends an illegal url so that a user may get in trouble if the honeservers logs are checked. I agree it isn't the worst privacy flaw but I think it is worth letting clients make the choice and provide users with options rather than writing this section like it is expected.

Author: > The top comment suggests to use "may" instead of "it is expected", which soru will incorporate into this MSC

> use "may" instead of "it is expected"

may means the clients still have a choice.

this MSC does not add any new locations where clients should display url previews, it is about limiting it in some messages. "**Disable** URL Previews".

Also there is a reason element disables URL previews in e2ee rooms by default.

Reviewer 3: I like that. I think it makes sense to focus on disallowing certain previews in this MSC and not sound like we are requiring any previews where they weren't required previous.

> present in the body it is expected to not render a preview for that URL.

Reviewer: ```suggestion
present in the body clients SHOULD NOT render a preview for that URL.
```

Author: Nope, we had this discussion in the other PR. They could e.g. show a preview only on hovering over the url or something, thus "SHOULD NOT" is a tad too strong wording, soru thinks.

Reviewer: AIUI, `url_previews` having URLs not present in the body is not considered appropriate use in the first place. Did I misunderstand? I'd really prefer `url_previews` to only contain a subset of URLs from the body but not other URLs.

Author: Aaaaah, right, yes, sorry, just thought this was on a different block of the MSC, didn't look too closely

> present in the body clients SHOULD NOT render a preview for that URL.

Reviewer 5: in the body or formatted body?

Reviewer 5: (answering this may lead to further questions and concerns)

Author: since all links in the formatted_body *should* be present in the body and vice-versa (as the body is a plain-text fallback) this should only matter for those messages that actively disobey that.

That being said, it seem to make most sense to use the links in whichever the client uses to render: if it renders formatted stuffs the formatted_body, if it renders plaintext the body

Reviewer 5: right, so the concern here would be that you can make a very convincing phishing attack: if it uses URLs only from `body`, your formatted body can be `<a href="https://scam.com">https://example.org</a>` with `url_previews` for `https://example.org`. Likewise, if it only looks at formatted bodies then the same attack can be applied.

Currently at least, clients will (or should) be showing a URL preview for the `href` rather than the text.

Reviewer: If a client uses `body` (including URLs in it, naturally) to render the message then `formatted_body` is irrelevant, whichever URLs it contains, and vice versa. `url_previews` is only used as a whitelist for URLs in whichever body the client renders. I don't quite see the surface for a phishing attack here: what you see is what you get, with a caveat that an unsuppressed URL preview for a phishy URL in `formatted_body` may help revealing the attack (at the very least, it won't conceal the attack), whereas if an attacker suppreses the URL preview then it's up to the user to check where that URL leads to before following it.

Reviewer 5: Sure, it's the responsibility of the user to check links before they click but in practice we shouldn't be designing into the protocol a requirement for people to do that. 

The attack is quite literally:
```json
{
  "body": "https://example.org",
  "format": "org.matrix.custom.html",
  "formatted_body": "<a href="https://phishing.example.org">https://example.org</a>",
  "url_previews": ["https://example.org"]
}
```

Which clients that don't read the spec to the letter would render as:
> https://example.org
> > Example Domain
> > This is an example domain.

Instead, we could be better at protecting users from badly written clients and from having to be more responsible than us.

Author: The fix for that is equally as trivial: As the client uses formatted_body to figure out URL previews, it is expected to figure out that the stuff in the href of the `<a>` tag should give the preview - not its content

Reviewer 3: Maybe it would be more clear if it was specified like the following:

> A client may chose to provide URL previews for the user. For each link the client finds in the message the client must consult the `url_previews` attribute. If `url_previews` is absent all URLs are allowed to be previewed, otherwise if the attribute is a list contacting the link URL, the URL is present for previewing.

So it is clear that the client must not read `url_previews` and search for the URL in the body, but instead if in the process of rendering the body it finds a link it should consult `url_previews`. In your example the client would never consider `https://example.org` a link and therefore would never be eligible for a preview whether or not `url_previews` was absent or `["https://example.org"]`

The consequence is that putting arbitrary URLs into `url_previews` accomplishes nothing unless the URL is actually going to be treated as a link by the client.

### Examples
No key added, preview to all URLs:
```json
{
	"msgtype": "m.text",
	"body": "Wow, check out this song! https://www.youtube.com/watch?v=oHg5SJYRHA0"
}
```

Empty array for `url_previews`, thus no URL preview at all:
```json
{
	"msgtype": "m.text",
	"format": "or.matrix.custom.html",
	"body": "Hey, look at what I did! [MSC2385: Disable URL Previews](https://github.com/matrix-org/matrix-doc/pull/2385)",
	"formatted_body": "Hey, look at what I did! <a src=\"https://github.com/matrix-org/matrix-doc/pull/2385\">MSC2385: Disable URL Previews</a>",
	"url_previews": []
}
```

Array of URLs to preview, only those in the array are previewed:
```json
{
	"msgtype": "m.text",
	"body": "https://example.com https://matrix.org",

> 	"url_previews": ["https://example.com"]

Reviewer 4: Use an list of 0-based indices, which would act as a selector that would enable previews of selected links (bare URIs, `<a/>` and `<link/>` count, `<url/>` never gets a preview).

Reviewer 6: Different clients may count URLs differently.  e.g. in the future, some clients may try to get previews for, say `gemini:` links, while others may have no idea what `gemini:` is.  So if someone sends a message of "gemini://example.com/ https://matrix.org", is https://matrix.org URL # 0 or # 1?

Reviewer 4: > Different clients may count URLs differently.

Unpreviewable URLs would also be counted in terms of indexing, making the indexes same for everyone.

Reviewer 7: So you'd need to spec the grammar of what constitutes a previewable URL so all clients can implement it consistently. Element Web for example uses soapbox/linkifyjs which doesn't support custom protocols so relying on that wouldn't be good for your proposal

Reviewer 4: Have you read my comment? It says unpreviewable URIs still count for indexing, so IETF RFC 3986 would apply. If you select an unpreviewable URI, you will not get preview of that particular URI only. So, it will fail very gracefully.

Reviewer 6: Some clients may treat "www.matrix.org" as being equivalent to https://www.matrix.org even though it's not a full URL and want to get a preview for that.  Some clients may wish to treat something like "https://matrix.org/,https://element.io/" or "(https://matrix.org/)https://element.io" as two URLs even though RFC 3986 says they're both valid as being a single URL.

Basically, if you want to go this route, you need to introduce a robust mechanism that all clients *must* follow (and inevitably some clients will implement incorrectly), and you may end up with edge cases that don't provide a good UX.

Reviewer 4: > edge cases that don't provide a good UX.

The client may warn the user in cases like that.

Reviewer 6: > The client may warn the user in cases like that.

Which client, when should they warn, and under what conditions?  You're making vague statements, but your proposal needs to be backed up by concrete, specific ideas.  The devil is in the details.

In any event, warnings to the user do not solve bad UX.  Warnings *are* bad UX.

Your proposal:

- makes an assertion about what should be done without giving any reasoning about why it's any better
- requires all clients to replicate the exact same behaviour
- is brittle -- if a client does not parse URLs in exactly the same way, if it deviates in the slightest way, it might not just produce a slightly incorrect result, it could end up with the exact opposite result from what was intended: previews for exactly the URLs that were not intended to be previewed, and no previews for the URLs that were intended to be previewed
- does not give clients any leeway to try to improve UX by accounting for what a person might mean by what they wrote, such as allowing them to just write "www.matrix.org" instead of the full URL.  Thus, not only are you dictating client behaviour, you are also trying to dictate user behaviour, which is much harder.  Designing something that will only work if the user does things in a specific way is a Bad Idea :tm:

Reviewer 4: > Which client, when should they warn, and under what conditions?

That warning would be an optional diagnostic, hence those conditions may be left implementation-defined. For example, in case of a bot, that may even log the not-so-well-formed-resource-identifier in question, or provide something like a highlight or a pop-up in case of an interactive client. Optional diagnostics are not alien to us. You know, suppressing warnings, or making them fatal.

Reviewer 6: > That warning would be an optional diagnostic, hence those conditions may be left implementation-defined. For example, in case of a bot, that may even log the not-so-well-formed-resource-identifier in question, or in case of an interactive client, provide something like a highlight or a pop-up. Optional diagnostics are not alien to us. You know, suppressing compiler warnings, or making them fatal (`-Wall -Werror` anyone?)

Again, not good UX

I'm going to resolve this conversation as I don't think that the change requested here requires a greater level of detail than what I think fits well in a comment thread.  If you really want to continue to pursue this, I'd suggest opening a new MSC *that includes the necessary detail*.

}
```

## Potential issues
URLs that should have a preview need to be sent twice - once in the body, once in `url_previews`. As

> this feature is meant for bots, normal human clients would probably omit `url_previews` alltogether,

Reviewer 6: ```suggestion
this feature is meant for bots, normal human clients would probably omit `url_previews` altogether,
```
... because English is weird.

> resulting in all URLs sent from them having a preview.

Reviewer 3: I see no reason a client can't display a list of links that it detects and allow the user to disable previews for some or all of them.

Reviewer 6: I must have missed this paragraph earlier.  I agree that I don't see any reason why this should be intended mainly for bots and not for humans.  There have been several occasions where I've typed a link and not wanted to have a preview because it causes extra noise (not just for rickrolling).

Reviewer 8: I, as a human, would have use for this every once in a while. Implementation examples of UI for this have been mentioned earlier (facebook, ...). I don't know if it should be added to the MSC, but since the point is to not send certain URLs to the HS, the sending UX should be constructed in such a way that, if at all, it is the client which generates the preview so that you can then decide to include it or not. The reasoning being that the sending user has (99% likely) already visited that URL and, if privacy is a concern, taken appropriate measures.

> URLs in `url_previews` and the body need to be an exact slash, which could be confusing with trailing

Reviewer: ```suggestion
URLs in `url_previews` and in the body need to match exactly, which could be confusing with trailing
```

slashes (`/`).

## Alternatives
There could be an attribute added to the `<a>` tag which disables URL previews. See
[MSC2376](https://github.com/matrix-org/matrix-doc/pull/2376).

Instead of a whitelist there could also be a blacklist of URLs. However, as the likely scenario is a
bot wanting to surpress all the previews for its URLs sent it seems like a whitelist is more appropriate

> here.

Reviewer 3: A whitelist also nicely prevents clients from parsing things that were not intended to be URLs and trying to preview them. Should mostly be relevant for `text` as `html` has a precise notion of what a link is but still nice to have.
